simulator/bypass/data_bypass.h

```diff
@@ -0,0 +1,172 @@
+        };
+
+        class BypassCommand
+        {
+            public:
+                BypassCommand( RegisterStage bypassing_stage, RegNum register_num)
+                    : bypassing_stage( bypassing_stage)
+                    , register_num( register_num)
+                { }
+
+                auto get_bypassing_stage() const { return bypassing_stage; }
+                auto get_register_num() const { return register_num; }    
+
+            private:
+                const RegisterStage bypassing_stage;
+                const RegNum register_num;
+        };
+
+    private:
+        struct RegisterInfo
+        {
+            RegisterStage current_stage = RegisterStage::in_RF();
+            RegisterStage ready_stage = RegisterStage::in_RF();
+            bool is_bypassible = false;
+            Cycle first_execute_stage_cycle = 0_Cl;
+        };
+
+
+        class Scoreboard
+        {
+            public:
+                RegisterInfo& get_entry( RegNum num)
+                {
+                    return array.at( static_cast<std::size_t>( num));
+                }
+
+                const RegisterInfo& get_entry( RegNum num) const
+                {
+                    return array.at( static_cast<std::size_t>( num));
+                }
+
+            private:
+                std::array<RegisterInfo, REG_NUM_MAX> array = {};
+        };
+
+        auto get_current_stage( RegNum num) const
+        {
+            return scoreboard.get_entry( num).current_stage;
+        }
+
+        void update_register_info( const MIPSInstr& instr, RegNum num,
+                                   const Cycle& cycle,
+                                   RegisterStage new_dst_stage);
+
+        void set_initial_state_to_reg_info( RegNum num)
+        {
+            auto& entry = scoreboard.get_entry( num);
+            entry.current_stage = RegisterStage::in_RF();
+            entry.is_bypassible = false; 
+        }
+
+        Scoreboard scoreboard = {};
+
+        /* gives the information whether bypassed data should be transformed
+           when bypassing is needed for HI register
+        */ 
+        bool is_HI_master_DIVMULT = false;
+
+    public:
+        auto is_in_RF( RegNum num) const
+        {
+            return scoreboard.get_entry( num).current_stage == RegisterStage::in_RF();
+        }
+
+        auto is_bypassible( RegNum num) const
+        {
+            return scoreboard.get_entry( num).is_bypassible;
+        }
+
+        auto is_stall( const MIPSInstr& instr) const
+        {
+            const auto src1_reg_num = instr.get_src1_num();
+            const auto src2_reg_num = instr.get_src2_num();
+
+            return ( (!is_in_RF( src1_reg_num) && !is_bypassible( src1_reg_num)) ||
+                     (!is_in_RF( src2_reg_num) && !is_bypassible( src2_reg_num)) );
+        }
+
+        auto get_bypass_command( RegNum num) const
+        {
+            return BypassCommand( get_current_stage( num), num);
+        }
+
```

Comment:
That class is very complicated and requires comments for each method:
- what parameters are
- what is expected and unexpected
- what return value is